sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -198,10 +198,12 @@ class Analyzer(
 
       g.bitmasks.foreach { bitmask =>
         // get the non selected grouping attributes according to the bit mask
-        val nonSelectedGroupExprSet = buildNonSelectExprSet(bitmask, g.groupByExprs)
+        val nonSelectedGroupExprs = buildNonSelectExprs(bitmask, g.groupByExprs)
 
         val substitution = (g.child.output :+ g.gid).map(expr => expr transformDown {
-          case x: Expression if nonSelectedGroupExprSet.contains(x) =>
+          case x: Expression
+            if nonSelectedGroupExprs.find(
+                 ExpressionEquality(_) == ExpressionEquality(x)).isDefined =>
```

Comment:
How about we add `semanticEquals` method to `Expression` and implement it with `equals` as default? Then we can override `semanticEquals` for `AttributeReference` and just write `_.semanticEquals(x)` here.